### PR TITLE
Add MSRV options to `unnested_or_patterns`

### LINKED ISSUE #6953: unnested_or_patterns does not respect clippy.toml msrv
Repro:

```toml
# .clippy.toml

msrv = "1.31.0"
```

```rust
// src/main.rs

#![deny(clippy::pedantic)]

enum E {
    A,
    #[allow(dead_code)]
    B,
}

fn main() {
    match E::A {
        e @ E::A | e @ E::B => {
            let _ = e;
        }
    }

    match E::A {
        e @ (E::A | E::B) => {
            let _ = e;
        }
    }
}
```

`cargo clippy` emits a unnested_or_patterns lint, despite that syntax not being valid on Rust 1.31.0.

```console
error: unnested or-patterns
  --> src/main.rs:11:9
   |
11 |         e @ E::A | e @ E::B => {
   |         ^^^^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![deny(clippy::pedantic)]
   |         ^^^^^^^^^^^^^^^^
   = note: `#[deny(clippy::unnested_or_patterns)]` implied by `#[deny(clippy::pedantic)]`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnested_or_patterns
help: nest the patterns
   |
11 |         e @ (E::A | E::B) => {
   |         ^^^^^^^^^^^^^^^^^
```

The suggested replacement on Rust 1.31.0:

```console
$ cargo +1.31.0 check
                                             
error: expected one of `)`, `,`, or `::`, found `|`                                       
  --> src/main.rs:11:19                                                                   
   |                                                                                      
11 |         e @ (E::A | E::B) => {                                                       
   |                   ^ expected one of `)`, `,`, or `::` here
```

<br>

### Meta

- `cargo clippy -V`: clippy 0.1.52 (5d04957 2021-03-22)
- `rustc -Vv`:
  ```
  rustc 1.53.0-nightly (5d04957a4 2021-03-22)
  binary: rustc
  commit-hash: 5d04957a4b4714f71d38326fc96a0b0ef6dc5800
  commit-date: 2021-03-22
  host: x86_64-unknown-linux-gnu
  release: 1.53.0-nightly
  LLVM version: 12.0.0
  ```

### PR DESCRIPTION
**Pull request by flip1995**
changelog: [`unnested_or_patterns`] can now be configured with the `msrv` config/attribute.

Fixes #6953

### CODE REVIEW

**Review by xFrednet — APPROVED** [review]
The documentation updates LGTM. The code as well, especially with the docs, but it might be good if someone who has worked with msrv double-checks the changes :upside_down_face:

**Comment by flip1995** [comment]
The only one assigned to PRs that really dealt with the MSRV feature was me :smile:

File: clippy_lints/src/unnested_or_patterns.rs [hunk 1]
```diff
@@ -50,26 +54,54 @@ declare_clippy_lint! {
...
+        if !meets_msrv(self.msrv.as_ref(), &UNNESTED_OR_PATTERNS_MSRV) {
+            return;
+        }
         if let ast::ExprKind::Let(pat, _) = &e.kind {
             lint_unnested_or_patterns(cx, pat);
         }
     }
 
     fn check_param(&mut self, cx: &EarlyContext<'_>, p: &ast::Param) {
+        if !meets_msrv(self.msrv.as_ref(), &UNNESTED_OR_PATTERNS_MSRV) {
```
Comment (llogiq) [inline clippy_lints/src/unnested_or_patterns.rs:91]: I think if we put the lint call in the `if` clause, we can remove the condition negation.

File: clippy_lints/src/unnested_or_patterns.rs [hunk 2]
```diff
@@ -50,26 +54,54 @@ declare_clippy_lint! {
...
+            return;
+        }
         lint_unnested_or_patterns(cx, &p.pat);
     }
 
     fn check_local(&mut self, cx: &EarlyContext<'_>, l: &ast::Local) {
+        if !meets_msrv(self.msrv.as_ref(), &UNNESTED_OR_PATTERNS_MSRV) {
```
Comment (llogiq) [inline clippy_lints/src/unnested_or_patterns.rs:98]: Same here

**Review by llogiq — COMMENTED** [review]
Just a small nit, otherwise LGTM.

**Comment by flip1995** [comment]
I just moved everything in the `if` block for consistency :+1:

**Comment by llogiq** [comment]
👍 @bors r+

**Pull request merged**